### PR TITLE
feat!: per-stream persistence to disk

### PR DESCRIPTION
**Pull request by de-sh**
Closes #

### Changes

BREAKING: persistence is now configurable only at the stream level, i.e. users need to define if a stream is persistable at the stream config level.
Example config where the gps stream is persisted to `/tmp/uplink/gps`, with 3 files allowed, each being maximum of 1MB in size:
```
[streams.gps]
topic = "/tenants/{tenant_id}/devices/{device_id}/events/gps/jsonarray"
persistence = { path = "/tmp/uplink/gps", max_file_size = 1048576, max_file_count = 3 }
```

### Why?

Users would want to opt-in to persistence on a per-stream basis.

### Trials Performed

Setup: uplink with above example config, connected to simulator which is generating data on the streams: `imu`, `gps`, `motor`, etc. where we have configured `gps` and `motor` to both persist, whereas `imu` is configured to not do so, we have also configured `imu` and `motor` to compress the batches of data received over the stream.
1. Stream configured without persistence.
Observations: default size of in-memory write buffer is 100MB, buffer on overflow is cleared and hence lost.
2. Stream configured with persistence.
Observations: in-memory write buffer on overflow is flushed onto disk in directory defined in stream config.
3. Stream configured without persistence and compression(topic in config will differ from what is received over data collector channel).
Observations: in-memory write buffer on overflow is flushed onto disk in directory defined in stream config, even though the topic on stream is different from what is defined in config.
```

/tmp/uplink
├── gps
│   ├── backup@1
│   ├── backup@10
│   ├── backup@2
│   ├── backup@3
│   ├── backup@4
│   └── backup@5
│   └── backup@6
│   └── backup@7
│   └── backup@8
│   └── backup@9
└── motor
    ├── backup@37
    ├── backup@38
    └── backup@39

2 directories, 9 files
```
In above example it can be observed that with `config.toml`, uplink writes max 10 files for gps backlog and 3 for motor.

[video explainer of process used to verify working](https://drive.google.com/file/d/1lyQAvSX7MegZu6DLDLNsTfYjVJmBKnWc/view?usp=sharing)

### CODE REVIEW

File: configs/config.toml [hunk 1]
```diff
@@ -80,18 +65,47 @@ max_file_count = 3
...
+# - path(optional): Path to directory for storing backlog in files, shouldn't contain anything else.
+#   Please ensure the location is unique for each stream to ensure there is no clash in files. 
+# - max_file_count(optional, defaults to 3): Maximum number of persistence files allowed on disk.
+#
+# NOTE: Persitence is an optional feature that is disabled by default, i.e. if not inlcuded in the 
+# configuration, we use transient(in-memory) storage to handle network downtime only.
+[streams.gps]
+topic = "/tenants/{tenant_id}/devices/{device_id}/events/gps/jsonarray"
+buf_size = 10
+persistence = { path = "/tmp/uplink/gps", max_file_size = 1048576, max_file_count = 10 }
```
Comment (tekjar) [inline configs/config.toml:98]: This will be problematic if same path is used in other stream's config. We should only specify base path here. Serializer init should create folder with `stream` name in base path

File: configs/config.toml [hunk 2]
```diff
@@ -80,18 +65,47 @@ max_file_count = 3
...
+
+# NOTE: While it is possible to configure persistence to be disabled, including only max_file_count
+# without path causes non-persistence. Configuring only the persistence path is allowed and the other
+# config values will be default. Configuring only max_file_size is also allowed, without persistance.
+[streams.motor]
+topic = "/tenants/{tenant_id}/devices/{device_id}/events/motor/jsonarray"
+buf_size = 50
+compression = "Lz4"
```
Comment (tekjar) [inline configs/config.toml:106]: Convert to small case?

File: configs/config.toml [hunk 3]
```diff
@@ -80,18 +65,47 @@ max_file_count = 3
...
+[streams.gps]
+topic = "/tenants/{tenant_id}/devices/{device_id}/events/gps/jsonarray"
+buf_size = 10
+persistence = { path = "/tmp/uplink/gps", max_file_size = 1048576, max_file_count = 10 }
+
+# NOTE: While it is possible to configure persistence to be disabled, including only max_file_count
+# without path causes non-persistence. Configuring only the persistence path is allowed and the other
+# config values will be default. Configuring only max_file_size is also allowed, without persistance.
+[streams.motor]
+topic = "/tenants/{tenant_id}/devices/{device_id}/events/motor/jsonarray"
```
Comment (tekjar) [inline configs/config.toml:104]: We should make `compression` in topic manual just like `serialization`. Topic should be changed to 

```
/tenants/{tenant_id}/devices/{device_id}/events/motor/jsonarray/lz4
```

File: uplink/src/base/serializer/mod.rs [hunk 4]
```diff
@@ -224,23 +263,24 @@ impl<C: MqttClient> Serializer<C> {
...
         // Reactlabs setup processes logs generated by uplink
         info!("Switching to slow eventloop mode!!");
         self.metrics.set_mode("slow");
 
         // Note: self.client.publish() is executing code before await point
         // in publish method every time. Verify this behaviour later
-        let payload = &publish.payload[..];
-        let publish = self.client.publish(&publish.topic, QoS::AtLeastOnce, false, payload);
+        let payload = Bytes::copy_from_slice(&publish.payload[..]);
+        let publish = send_publish(self.client.clone(), publish.topic, payload);
```
Comment (tekjar) [inline uplink/src/base/serializer/mod.rs:274]: We shouldn't clone clients without out understanding what channel `tx` in `client` does on clone. In some implementations, it can result in indefinite capacity increase

**Comment by tekjar** [comment]
This change should be chaos tested with toxiproxy covering all persistence related tests in the QA document

File: uplink/src/lib.rs [hunk 5]
```diff
@@ -195,9 +191,8 @@ pub mod config {
                     topic: format!(
                         "/tenants/{tenant_id}/devices/{device_id}/events/{stream_name}/jsonarray"
                     ),
-                    buf_size: config.system_stats.stream_size.unwrap_or(100),
```
Comment (tekjar) [inline uplink/src/lib.rs:198]: Are these changes necessary in this PR?

File: uplink/src/base/mod.rs [hunk 6]
```diff
@@ -50,6 +59,26 @@ pub struct StreamConfig {
...
+    pub fn topic(&self) -> String {
+        self.topic.to_owned() + self.compression.suffix()
+    }
+}
+
+impl Default for StreamConfig {
+    fn default() -> Self {
+        Self {
+            topic: "".to_string(),
+            buf_size: 1,
```
[the remark this answers is not in the document]
Comment (de-sh) [reply uplink/src/base/mod.rs:76]: Would it be a good idea to rename `MAX_BUFFER_SIZE` as `DEFAULT_BUFFER_SIZE`?

File: uplink/src/base/serializer/mod.rs [hunk 7]
```diff
@@ -184,26 +187,61 @@ impl<C: MqttClient> Serializer<C> {
...
         })
     }
 
+    fn select_storage(&mut self, topic: &str) -> &mut Storage {
+        self.storage_map
+            .entry(topic.to_owned())
+            .or_insert_with(|| Storage::new(default_file_size()))
+    }
+
+    fn next_storage(&mut self) -> Result<&mut Storage, Error> {
```
[the remark this answers is not in the document]
Comment (de-sh) [reply uplink/src/base/serializer/mod.rs:217]: The context is currently held completely inside of `Serializer` as the current idea revolves around storage per topic, we can create another module and struct to namespace this along the lines of a `storage_handler`, but that would abstract things more than necessary in my opinion, your call thought.

File: uplink/src/base/serializer/mod.rs (hunk 4 above)
Comment (de-sh) [reply uplink/src/base/serializer/mod.rs:274]: In the case of flume we are merely cloning the handle to the channel, [the channel is bounded anyways and will not increase allocation/capacity on clone](https://docs.rs/flume/0.10.14/flume/struct.Sender.html#method.clone).

File: configs/config.toml (hunk 3 above)
Comment (de-sh) [reply configs/config.toml:104]: I agree

File: configs/config.toml (hunk 1 above)
Comment (de-sh) [reply configs/config.toml:98]: This is exactly what I had called to ask you about, would the customer expect the behavior of creating a sub-directory or not.

File: configs/config.toml (hunk 2 above)
Comment (de-sh) [reply configs/config.toml:106]: serde will not recognize it in that case, `"Lz4"` and any other compression type that we include in the future will have to adhere to the same, looked around for a compatible solution, have not been able to find one. Please raise an issue if you feel this is a necessary change, else I feel it is best left untouched as there is no relation to the current PR beyond the config level.

File: uplink/src/base/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tekjar) [reply uplink/src/base/mod.rs:76]: We can hard code and take clean up in a separate task?

File: uplink/src/lib.rs (hunk 5 above)
Comment (tekjar) [reply uplink/src/lib.rs:198]: `unwrap_or` is changed to `MAX_BUFFER_SIZE`. Looks unrelated and no idea what that value is

File: uplink/src/base/serializer/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (de-sh) [reply uplink/src/base/serializer/mod.rs]: https://github.com/bytebeamio/uplink/pull/229/commits/894bd1f79b20bac5ec828ffd16e5a1ec0665a57a

File: configs/config.toml (hunk 3 above)
Comment (de-sh) [reply configs/config.toml:104]: Resolved in #230

**Comment by de-sh** [comment]
QA performed: 
- [x] Writes to disk on bad network when disk persistence is configured
- [x] Is any file larger than the configured file size. 
NOTE: Seems like files are larger by a very good margin for smaller sized files and not so much for larger ones, this might be caused by attached metadata.
```
$ ls -lh /tmp/uplink/gps # max_file_size = 1024
...
-rw-rw-r-- 1 devdutt devdutt 1.8K May 11 00:43 backup@0
-rw-rw-r-- 1 devdutt devdutt 1.8K May 11 00:43 backup@1
...
$ ls -lh /tmp/uplink/motor # max_file_size = 102400
...
-rw-rw-r-- 1 devdutt devdutt 101K May 11 00:44 backup@0
...
- [x] Applies memory and disk retention policies correctly (Files beyond max_file_count are being deleted)
- [x] On slow networks, uplink reads and writes from/to disk simultaneously (in catchup mode, which is triggered the moment a single ack is received in mqtt)
- [x] Slow mode, uplink writes to in-memory and network recovers before reaching max-file-size. lost = 0, disk_files = 0 in uplink_serializer_metrics.
- [x] Slow mode, uplink can't write to disk but max-file-size is breached. lost > 0, disk_files = 0 in uplink_serializer_metrics. (in-memory buffer is cleared, observed in the case of imu stream in slow mode)
- [x] Slow mode, uplink writes to disk and network recovers before reaching max-file-count. lost = 0, disk_files > 0 < max_file_count in uplink_serializer_metrics. (verified with gps stream with max_file_count = 10)
- [x] Slow mode, uplink writes to disk and network recovers after reaching max-file-count. lost > 0 in uplink_serializer_metrics (verified with motor, which counted up lost files)
- [x] When there are files in persistence directory at startup, does uplink read and forward them appropriately, while writing incoming data onto disk and waiting for the reading to be completed before writing the incoming data directly to network (this is observed best when uplink is connected through a slow proxy)
